LayoutTests/css3/filters/resources/filter-helpers.js

```diff
@@ -0,0 +1,210 @@
+
+function buildCSSFilteredBoxes(filterValues)
+{
+    for (const value of filterValues) {
+        const img = new Image();
+        img.src = 'resources/reference.png';
+        img.setAttribute('style', `filter: ${value}`);
+        document.body.appendChild(img);
+    }
+}
+
+function buildSVGFilteredBoxes(filterValues)
+{
+    const filterIDs = buildSVGFilters(filterValues);
+
+    for (const filterID of filterIDs) {
+        const img = new Image();
+        img.src = 'resources/reference.png';
+        img.setAttribute('style', `filter: url('#${filterID}');`);
+        document.body.appendChild(img);
+    }
+}
+
+const filterNameToFunctionMap = {
+    'brightness' : svgBrightnessComponent,
+    'contrast' : svgContrastComponent,
+    'grayscale' : svgGrayscaleComponent,
+    'sepia' : svgSepiaComponent,
+    'saturate' : svgSaturateComponent,
+    'hue-rotate' : svgHueRotateComponent,
+    'invert' : svgInvertComponent,
+    'opacity' : svgOpacityComponent,
+    'blur' : svgBlurComponent,
+    'drop-shadow' : svgDropShadowComponent,
+};
+
+function buildSVGFilters(filterValues)
+{
+    const svg = document.createElementNS('http://www.w3.org/2000/svg', 'svg');
+    let filterIDs = [];
+    
+    let filterIndex = 0;
+    for (const value of filterValues) {
+
+        let filterElement = document.createElementNS('http://www.w3.org/2000/svg', 'filter');
+        
+        const filterFunctionPattern = /[-a-zA-Z]+\(.+?\)/g; // Doesn't support nested parens.
+        const filterList = [...value.matchAll(filterFunctionPattern)];
+
+        for (const filterString of filterList) {
+            const filterPattern = /(.+)\((.*)\)/;
+            const matchResult = filterPattern.exec(filterString);
+            const filterType = matchResult[1];
+            const param = matchResult[2];
+
+            const filterCreationFunction = filterNameToFunctionMap[filterType];
+            if (!filterCreationFunction) {
+                console.log(`Failed to create filter of type ${filterType}`);
+                continue;
+            }
+
+            if (filterType === 'drop-shadow') {
+                // Because the SVG equivalent of drop-shadow involves multiple components,
+                // we can't support it in a list.
+                filterElement = filterCreationFunction(param);
```

Comment:
I think it is okay but we have to build the drop-shadow as a <filter> element not as a primitive element?